Add jerryscript-ext/autorelease.h with JERRYX_AR_VALUE_T

martijnthe opened the pull request:
As discussed/proposed here:
https://github.com/jerryscript-project/jerryscript/issues/1836

## TODOs:

- [x] Add documentation

martijnthe commented:
cc @jiangzidong @zherczeg @akosthekiss @LaszloLango

jiangzidong commented:
LGTM, please fix the coding-style issues.

LaszloLango commented on jerry-ext/autorelease/autorelease.c:
```diff
@@ -0,0 +1,26 @@
+void
+jerryx_autorelease (const jerry_value_t *value) /**< Pointer to the value to release */
+{
+	jerry_release_value(*value);
```
Please don't use tabs :)

martijnthe commented:
OK, fixed the code style issues.

martijnthe commented:
Any comments on the naming of the macros?
`JERRYX_VAL` and `JERRYX_UNUSED` don't really explain that they are "autoreleasing".
But adding `AUTO` or something to the name makes the macro pretty long...

Once the naming is decided, I'll add documentation.

LaszloLango commented:
I agree. The current names are not descriptive enough. I think adding the `AUTO` does not make it too long. Maybe `AR` would be a good acronym for autorelease.

martijnthe commented:
Maybe it's not too long indeed...

Both `JERRYX_AR_VAL` / `JERRYX_AUTO_VAL` are still shorter compared to `const jerry_value_t`, which is what I would otherwise write...

zherczeg commented on jerry-ext/include/jerryscript-ext/autorelease.impl.h:
```diff
@@ -0,0 +1,47 @@
+ * The GCC __cleanup__ function must take a pointer to the variable to clean up.
+ */
+void jerryx_autorelease (const jerry_value_t *value);
+#define __JERRYX_VAL_IMPL const jerry_value_t __attribute__ ((__cleanup__(jerryx_autorelease)))
```
Is `const` required here?

martijnthe replied:
Strictly speaking `const` is not required here, but leaving out `const` will make it hairier to use it because the compiler will only call `jerryx_autorelease` for the last value that was assigned.
With `const`, using the macro is a bit more fool-proof.

If you really need a mutable `jerry_value_t`, I think you're better off 1) just writing `jerry_value_t` or 2) restructuring the code to avoid the need for mutability :)

Makes sense?

akosthekiss commented on jerry-ext/include/jerryscript-ext/autorelease.h:
```diff
@@ -0,0 +1,60 @@
+ *   // of scope.
+ * }
+ */
+#define JERRYX_VAL __JERRYX_VAL_IMPL
```
let's not pollute the `jerryx_` "global" namespace. `JERRYX_VAL` sounds a bit like a catch-all term. `jerryx_auto_` `jerryx_ar_` `jerryx_autorelease_` would all be better. and one more note: this macro really acts like a type, it is the "smarter" variant of `jerry_value_t`. would the name `JERRYX_AR_VALUE_T` not align better with existing identifier names?

martijnthe replied:
sgtm!

akosthekiss commented on jerry-ext/include/jerryscript-ext/autorelease.impl.h:
```diff
@@ -0,0 +1,47 @@
+
+#define __JERRYX_NEXT_UNUSED_NAME __jerryx_unused_##__COUNTER__
+#define __JERRYX_UNUSED_IMPL __JERRYX_VAL_IMPL __attribute__ ((unused)) __JERRYX_NEXT_UNUSED_NAME
+#else /* !__GNUC__ */
```
if clang supports the attribute, we should have it in the extension from the start

martijnthe replied:
This already works with clang because clang defines `__GNUC__`.

akosthekiss replied:
Ah, OK then

akosthekiss commented on jerry-ext/include/jerryscript-ext/autorelease.h:
```diff
@@ -0,0 +1,60 @@
+#ifndef JERRYX_AUTORELEASE_H
+#define JERRYX_AUTORELEASE_H
+
+#include "autorelease.impl.h"
```
let's put the include within the `extern "C"` and drop the `extern "C"` from autorelease.impl.h (just like with arg.h / arg.impl.h)

akosthekiss commented on jerry-ext/autorelease/autorelease.c:
```diff
@@ -0,0 +1,26 @@
+#include "jerryscript.h"
+
+/**
+ * Calls jerry_release_value(*value).
```
(style) let's add that space before the parenthesis, even in the comment

martijnthe commented on tests/unit-ext/test-ext-autorelease.c:
```diff
@@ -0,0 +1,77 @@
+static void
+test_autorelease_unused_val (void)
+{
+  JERRYX_UNUSED = create_object ();
```
Good points. Removing `JERRYX_UNUSED`.

martijnthe commented:
OK, reviewed your comments.

Once the macro name is OK with everyone, I'll add some documentation to this PR.

martijnthe commented:
@LaszloLango @akosthekiss @zherczeg how does it look now? Is the naming OK?
If so, I'll add some docs. Thanks.

LaszloLango commented:
I'm OK with the current name.

zherczeg commented on jerry-ext/include/jerryscript-ext/autorelease.h:
```diff
@@ -0,0 +1,44 @@
+ *   // of scope.
+ * }
+ */
+#define JERRYX_AR_VALUE_T __JERRYX_AR_VALUE_T_IMPL
```
Why do we need to redefine the macro?

zherczeg replied:
Hm I would even consider a lower case name here although it would hurt the current current style. So this is just an idea.

Something like jerryx_autorel_val_t.

This is just an idea, not a request. I am curious about others opinion.

martijnthe replied:
> Why do we need to redefine the macro?

I put the implementation separately to keep this header cleaner and easier to consume (to avoid having the `#ifdef __GNUC__ ...` stuff here basically)

martijnthe replied:
> Something like jerryx_autorel_val_t.

Hmm personal opinion, but I'd rather leave macros always upper case, communicating that "something happens under the hood".

jiangzidong replied:
Upper case +1

zherczeg commented on jerry-ext/include/jerryscript-ext/autorelease.impl.h:
```diff
@@ -0,0 +1,33 @@
+ * Calls jerry_release_value (*value).
+ * The GCC __cleanup__ function must take a pointer to the variable to clean up.
+ */
+void jerryx_autorelease_cleanup (const jerry_value_t *value);
```
Do we need a pointer here? That forces the value onto the stack.

martijnthe replied:
Yeah, that's the way `__cleanup__` works:

> The function must take one parameter, a pointer to a type compatible with the variable.

https://gcc.gnu.org/onlinedocs/gcc/Common-Variable-Attributes.html

zherczeg replied:
Is it possible to inline the release function? That would solve the problem.

martijnthe replied:
Hah, apparently it's possible to inline the `__cleanup__` function. TIL :)
Thanks!

martijnthe commented:
OK, updated and added docs.
Hopefully the additions to update-webpage.sh work.
I don't couldn't figure out how to build & test the site locally...

@jiangzidong feel free to "commandeer" this PR when I'm away coming 2 weeks.

LaszloLango commented on docs/11.EXT-REFERENCE-AUTORELEASE.md:
```diff
@@ -0,0 +1,46 @@
+
+static void foo (bool enable)
+{
+    JERRYX_AR_VALUE_T bar = jerry_create_string (...);
```
Please indent with 2 spaces not 4.

LaszloLango commented on jerry-ext/include/jerryscript-ext/autorelease.h:
```diff
@@ -0,0 +1,54 @@
+ * Example usage:
+ * static void foo (bool enable)
+ * {
+ *    JERRYX_AR_VALUE_T bar = jerry_create_string (...);
```
ditto

LaszloLango commented on jerry-ext/include/jerryscript-ext/autorelease.h:
```diff
@@ -0,0 +1,54 @@
+ *
+ *       // ...
+ *
+ *       // jerry_release (baz) and jerry_release (bar) is called automatically before
```
comment in a comment :) It's a bit odd.

jiangzidong replied:
Is that OK? Because we use `//`, so I think people can understand it

LaszloLango replied:
sure, you can leave as is.

LaszloLango commented on docs/11.EXT-REFERENCE-AUTORELEASE.md:
```diff
@@ -0,0 +1,46 @@
+
+- [jerry_value_t](../api-reference/#jerry_value_t)
+- [jerry_acquire_value](../api-reference/#jerry_acquire_value)
+- [jerry_release_value](../api-reference/#jerry_release_value)
```
Are these links working correctly?

jiangzidong commented:
@LaszloLango Ha, after several attempts, I found the way to let one doc link to another doc's symbol
```
- [jerry_value_t](../docs/02.API-REFERENCE.md#jerry_value_t)
```

LaszloLango commented on docs/11.EXT-REFERENCE-AUTORELEASE.md:
```diff
@@ -0,0 +1,46 @@
+{
+  JERRYX_AR_VALUE_T bar = jerry_create_string (...);
+
+  if (enable) {
```
```c
if (enable)
{
```

LaszloLango commented:
@bzsolt Will these kind of references do work on the webpage too?

`[jerry_value_t](../docs/02.API-REFERENCE.md#jerry_value_t)`

jiangzidong commented:
@akosthekiss @bzsolt Could you please help review this patch?

bzsolt commented on docs/11.EXT-REFERENCE-AUTORELEASE.md:
```diff
@@ -0,0 +1,47 @@
+    ...
+   
+    /* 
+     * jerry_release (baz) and jerry_release (bar) is called automatically before
```
`jerry_release_value`

bzsolt commented on jerry-ext/include/jerryscript-ext/autorelease.h:
```diff
@@ -0,0 +1,54 @@
+ *
+ *     // ...
+ *
+ *     // jerry_release (baz) and jerry_release (bar) is called automatically before
```
`jerry_release_value`

bzsolt commented on tools/update-webpage.sh:
```diff
@@ -44,6 +45,7 @@ titles[$REFERENCE_COUNTING_MD]="Reference Counting"
 titles[$DEBUGGER_MD]="Debugger"
 titles[$CODING_STANDARDS_MD]="Coding Standards"
 titles[$EXT_REFERENCE_ARG_MD]="'Extension API: Argument Validation'"
+titles[$EXT_REFERENCE_AUTORELEASE_MD]="'Extension API: Autorelease values'"
```
Please start every word with a capital letter like the others: `Autorelease Values`

bzsolt commented on docs/11.EXT-REFERENCE-AUTORELEASE.md:
```diff
@@ -0,0 +1,47 @@
+automatically called when the variable goes out of scope.
+
+*Note*: The macro depends on compiler support. For GCC and LLVM/clang, the macro is implemented
+using the `__cleanup__` variable attribute. For other compilers, no support has been added yet. 
```
There are some trailing whitespace in this doc, please remove them.

akosthekiss commented on tools/update-webpage.sh:
```diff
@@ -44,6 +45,7 @@ titles[$REFERENCE_COUNTING_MD]="Reference Counting"
 titles[$DEBUGGER_MD]="Debugger"
 titles[$CODING_STANDARDS_MD]="Coding Standards"
 titles[$EXT_REFERENCE_ARG_MD]="'Extension API: Argument Validation'"
+titles[$EXT_REFERENCE_AUTORELEASE_MD]="'Extension API: Autorelease Values'"
```
I see no reason for listing the titles in a different order than the file names. This line should come last in this list.
